File: frontend/profiles/moda-infantil/kids-size.ts
/**
 * Tamanhos da grade de variantes do perfil moda_infantil (roupa de criança / varejo, camada 8.22) —
 * espelho 1:1 de src/main/java/com/meada/profiles/modainfantil/KidsSize.java.
 *
 * ESCAPADA da camada (vs lingerie): o eixo de tamanho aqui são FAIXAS ETÁRIAS (age bands), não as
 * letras PP..XGG da lingerie. O KidsSizeParityTest (backend) garante que os ids aqui e no enum Java
 * nunca divergem (o teste casa textualmente cada objeto `{ id: '...' }` deste arquivo).
 * id = string estável ASCII (persistida em moda_infantil_variants.size); label = rótulo exibido.
 */
export const KIDS_SIZES = [
  { id: 'RN', label: 'Recém-nascido' },
  { id: '0-3m', label: '0 a 3 meses' },
  { id: '3-6m', label: '3 a 6 meses' },
  { id: '6-9m', label: '6 a 9 meses' },
  { id: '9-12m', label: '9 a 12 meses' },
  { id: '1a', label: '1 ano' },
  { id: '2a', label: '2 anos' },
  { id: '3a', label: '3 anos' },
  { id: '4a', label: '4 anos' },
  { id: '6a', label: '6 anos' },
  { id: '8a', label: '8 anos' },
  { id: '10a', label: '10 anos' },
  { id: '12a', label: '12 anos' },
] as const

export type KidsSize = (typeof KIDS_SIZES)[number]
export type KidsSizeId = KidsSize['id']

/** Rótulo de um tamanho/faixa etária (fallback: o próprio id se desconhecido). */
export function sizeLabel(id: string): string {
  return KIDS_SIZES.find((s) => s.id === id)?.label ?? id
}

File: src/main/java/com/meada/profiles/modainfantil/KidsSize.java
package com.meada.profiles.modainfantil;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 * Eixo de TAMANHO das variantes do perfil moda_infantil (camada 8.22), MATERIALIZADO — espelho 1:1 de
 * {@code frontend/profiles/moda-infantil/kids-size.ts} ({@code KidsSizeParityTest}). Mesmo padrão de
 * enum hardcoded+parity do {@link com.meada.profiles.lingerie.LingerieSize}.
 *
 * <p>⭐ ESCAPADA da camada (vs lingerie, cujos tamanhos eram PP..XGG): aqui o eixo de tamanho são
 * FAIXAS ETÁRIAS (age bands: RN, 0-3m, 3-6m, ..., 10a, 12a). A outra dimensão da variante,
 * {@code color}, continua TEXTO LIVRE — não há enum de cor. O size validado aqui é a defesa app-level
 * ao cadastrar uma variante; a CHECK do banco em {@code moda_infantil_variants.size} é apenas de
 * comprimento (1..20), então a validação real do conjunto de tamanhos é deste enum.
 *
 * <p>⭐ {@link #suggestForAgeMonths(int)}: dado a idade da criança em meses, sugere a melhor faixa
 * etária — a IA usa pra sugerir um tamanho a partir da idade informada pela cliente. Cada banda tem
 * um teto exclusivo em meses; a banda escolhida é a primeira cujo teto > idade (a última, 12a, é o
 * topo e acolhe qualquer idade maior).
 */
public enum KidsSize {
    RN("RN", 1),        // recém-nascido: < 1 mês
    M0_3("0-3m", 3),    // 0 a 3 meses: < 3 meses
    M3_6("3-6m", 6),    // 3 a 6 meses: < 6 meses
    M6_9("6-9m", 9),    // 6 a 9 meses: < 9 meses
    M9_12("9-12m", 12), // 9 a 12 meses: < 12 meses
    A1("1a", 24),       // 1 ano: < 24 meses
    A2("2a", 36),       // 2 anos: < 36 meses
    A3("3a", 48),       // 3 anos: < 48 meses
    A4("4a", 72),       // 4 anos: < 72 meses (até ~5 anos e meio)
    A6("6a", 96),       // 6 anos: < 96 meses
    A8("8a", 120),      // 8 anos: < 120 meses
    A10("10a", 144),    // 10 anos: < 144 meses
    A12("12a", Integer.MAX_VALUE);  // 12 anos: topo (qualquer idade maior)

    private final String id;
    private final int upperBoundMonthsExclusive;

    KidsSize(String id, int upperBoundMonthsExclusive) {
        this.id = id;
        this.upperBoundMonthsExclusive = upperBoundMonthsExclusive;
    }

    /** O id é também o rótulo curto exibido (RN, 0-3m, 1a, ...). */
    public String id() {
        return id;
    }

    public String label() {
        return id;
    }

    /** Resolve um tamanho/faixa etária pelo id estável. Optional vazio se inválido/null. */
    public static Optional<KidsSize> fromId(String id) {
        if (id == null) {
            return Optional.empty();
        }
        return Arrays.stream(values()).filter(s -> s.id.equals(id)).findFirst();
    }

    /** Todos os tamanhos/faixas, na ordem de declaração (ordem etária crescente = ordem de exibição). */
    public static List<KidsSize> allActive() {
        return List.of(values());
    }

    /**
     * ⭐ Sugere a melhor faixa etária para uma criança de {@code months} meses. Idade negativa é
     * tratada como 0 (RN). Percorre as bandas na ordem etária e devolve a primeira cujo teto
     * (exclusivo) seja maior que a idade — ex.: 0→RN, 2→0-3m, 5→3-6m, 18→1a, 24→2a, 120→10a. Acima da
     * última banda definida, devolve A12 (12a), o topo.
     */
    public static KidsSize suggestForAgeMonths(int months) {
        int m = Math.max(months, 0);
        for (KidsSize size : values()) {
            if (m < size.upperBoundMonthsExclusive) {
                return size;
            }
        }
        return A12;   // inalcançável (A12 tem teto MAX_VALUE), mas mantém o compilador/contrato felizes.
    }
}
